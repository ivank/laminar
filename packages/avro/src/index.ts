/**
 * @packageDocumentation
 * @module @laminar/avro
 */
export { AvroDecimal, bufferToDecimal, decimalToBuffer, to64Bit } from './avro-decimal';
export { AvroTimestampMillis } from './avro-timestamp-millis';
export { AvroEpochDays } from './avro-epoch-days';
