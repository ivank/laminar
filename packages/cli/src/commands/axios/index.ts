import { Command, createCommand, Option } from 'commander';
import fs from 'fs';
import { printDocument } from '@laminar/ts-compose';
import { convertOapi } from './convert';
import chalk from 'chalk';
import { dirname } from 'path';
import { Logger } from '../../types';
import { toContext, concatStreamToString, parseSchema } from '../../helpers';

export const processFile = async (file: string): Promise<string> => {
  const { context, value } = await toContext(file);
  return printDocument(convertOapi(context, value));
};

export const axiosCommand = (logger: Logger = console): Command =>
  createCommand('axios')
    .description('Convert openapi schemas to typescript types for axios')
    .option('-f, --file <file>', 'Filename to process, uses STDIN if not specified')
    .addOption(new Option('-t, --stdin-type <type>', 'Type of STDIN data').default('json').choices(['json', 'yaml']))
    .option('-o, --output <output>', 'File to output to, uses STDOUT if not specified')
    .action(async ({ file, output, stdinType }) => {
      const input = file ? file : parseSchema(stdinType, await concatStreamToString(process.stdin));
      const result = await processFile(input);

      if (output) {
        fs.mkdirSync(dirname(output), { recursive: true });
        logger.info(`OpanAPI Schema ${chalk.green(file ? file : 'STDIN')} -> ${chalk.yellow(output)} axios types`);
        fs.writeFileSync(output, result);
      } else {
        process.stdout.write(result);
      }
    });
