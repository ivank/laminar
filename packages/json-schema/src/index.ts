/**
 * @packageDocumentation
 * @module @laminar/json-schema
 */

export { ValidationError } from './ValidationError';
export { ResolveError } from './ResolveError';
export {
  InvalidCode,
  Invalid,
  Messages,
  Options,
  Validator,
  Validation,
  error,
  empty,
  onlyErrors,
  errors,
  evaluateItem,
  evaluateProperty,
  skipRest,
  hasErrors,
  hasEvaluatedItem,
  hasEvaluatedProperty,
  combine,
  combineErrors,
  childOptions,
  validateSchema,
} from './validation';
export { Schema, JsonSchema, PrimitiveType } from './schema';
export { toMessages, toMessage, messages } from './messages';
export {
  isCompiled,
  withinContext,
  toSchemaObject,
  compile,
  ResolvedSchema,
  resolve,
  resolveFile,
  RefMap,
  FileContext,
} from './resolve';
export {
  validate,
  validateCompiled,
  ensureValid,
  ensureValidCompiled,
  ValidateOptions,
  ValidateSchemaOptions,
  ValidateCompiledOptions,
  Result,
  ResultError,
  ResultSuccess,
  Drafts,
  FormatErrors,
} from './validate';
export { CoerceCompiledOptions, CoerceSchemaOptions, CoerceOptions, coerce, coerceCompiled } from './coerce';
export { coerceSchema, coercers, CoercerOptions, Coercer } from './coercion';
