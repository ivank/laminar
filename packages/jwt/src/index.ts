/**
 * @packageDocumentation
 * @module @laminar/jwt
 */

export { RequestAuthInfo, RequestSign, User, Session, JWTData, JWTSign, JWTVerify, VerifyJWTData } from './types';
export { jwkPublicKey, JWKHandlerOptions } from './jwk';
export {
  authMiddleware,
  createSession,
  verifyBearer,
  jwtSecurityResolver,
  verifyToken,
  createSessionMiddleware,
  verifyJWT,
} from './jwt';
export {
  verifyKeycloack,
  keycloakAuthMiddleware,
  keycloakJwtSecurityResolver,
  JWTDataKeycloak,
  KeycloakOptions,
} from './keycloak';
