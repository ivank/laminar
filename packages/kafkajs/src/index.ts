/**
 * @packageDocumentation
 * @module @laminarjs/kafkajs
 */
export { KafkaConsumerService } from './kafka-consumer.service';
export { KafkaConsumerOptionalService } from './kafka-consumer-optional.service';
export {
  KafkaProducerService,
  RegisterSchemas,
  producerMiddleware,
  ProducerContext,
  registerSchemas,
  toProducerRecord,
  ToProducerRecordConfig,
  RegisterSchemasConfig,
} from './kafka-producer.service';
export { kafkaLogCreator, produce, Produce } from './helpers';
export {
  DecodedKafkaMessage,
  DecodedEachMessagePayload,
  DecodedBatch,
  DecodedEachBatchPayload,
  EachBatchConsumer,
  EachMessageConsumer,
  EncodedMessage,
  EncodedProducerRecord,
  SchemaRegistryConsumerRunConfig,
  EncodedSchemaProducerRecord,
  EachMessageeMiddleware,
  EachBatchMiddleware,
} from './types';
export { KafkaAdminService, CreateTopics } from './kafka-admin.service';
