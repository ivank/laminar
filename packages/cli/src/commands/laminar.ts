import commander from 'commander';
import { apiCommand } from './api';
import { axiosCommand } from './axios';
import { jsonSchemaCommand } from './json-shema';
import { convertCommand } from './avro';
import { Logger } from '../types';
import { fixturesCommand } from './fixtures';

export const laminarCommand = (output: Logger = console): commander.Command =>
  commander
    .createCommand('laminar')
    .version('0.15.0')
    .description(
      `Laminar CLI - a tool for converting OpenApi and Avro Schemas to appropriate typescript types

laminar api
  Convert openapi schema to types used by laminar api (openApiTyped middleware)

laminar axios
  Convert openapi schema to types used by axios

laminar json-schema
  Convert json schema to typescript types

laminar avro
  Convert avro to typescript types

laminar fixtures
  Load pg tables and generate @laminar/fixtures types from them

Example:
  laminar api --file openapi.yaml --output src/__generated__/openapi.yaml.ts
  curl http://example.com/openapi.yaml | laminar api --stdin-type yaml | prettier --stdin-filepath openapi.yamlts > src/__generated__/openapi.yaml.ts
  laminar api --file openapi.yaml --output src/__generated__/openapi.yaml.ts --watch
  laminar avro path/to/*.avsc --output-dir src/__generated__ --logical-type timestamp-millis=Date
  laminar fixtures postgres://example-admin:example-pass@localhost:5432/example --output src/__generated__/fixtures.ts
`,
    )
    .addCommand(apiCommand(output))
    .addCommand(axiosCommand(output))
    .addCommand(jsonSchemaCommand(output))
    .addCommand(convertCommand(output))
    .addCommand(fixturesCommand(output));
