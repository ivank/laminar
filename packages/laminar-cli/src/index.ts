import { ResolvedSchema } from '@ovotech/json-schema';
import { printDocument } from '@ovotech/ts-compose';
import { oas31 } from 'openapi3-ts';
import { convertOapi as apiConvert } from './commands/api/convert';
import { convertOapi as axiosConvert } from './commands/axios/convert';
import { convertSchema } from './convert-schema';

export { convertSchema } from './convert-schema';

export const apiContent = ({ schema, refs }: ResolvedSchema): string =>
  printDocument(
    apiConvert(
      { root: schema as oas31.OpenAPIObject & { [index: string]: unknown }, refs },
      schema as oas31.OpenAPIObject & { [index: string]: unknown },
    ),
  );

export const axiosContent = ({ schema, refs }: ResolvedSchema): string =>
  printDocument(
    axiosConvert(
      { root: schema as oas31.OpenAPIObject & { [index: string]: unknown }, refs },
      schema as oas31.OpenAPIObject & { [index: string]: unknown },
    ),
  );

export const schemaContent = ({ schema, refs }: ResolvedSchema): string =>
  printDocument(convertSchema({ root: schema as oas31.SchemaObject, refs }, schema));

export { toTypeScript as avroConvert } from './commands/avro/convert';
export { toTypeScript as fixturesConvert } from './commands/fixtures/convert';
export { toExternalContext as avroExternalConvert } from './commands/avro/convert';
