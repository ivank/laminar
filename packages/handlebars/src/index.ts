/**
 * @packageDocumentation
 * @module @laminar/handlebars
 */

import { Middleware } from '@laminar/laminar';
import { readdirSync, readFileSync, existsSync } from 'fs';
import { compile, RuntimeOptions, TemplateDelegate } from 'handlebars';
import { join, relative } from 'path';
import { statSync } from 'fs';

export type Cache = Map<string, { template: TemplateDelegate; mtime: Date }>;

/**
 * Cache types.
 * - preload: load all the templates ones into memory
 * - expiry: load partials when needed and keep them in cache, but check the file's mtime and reload template if changed
 * - none: do not cache templates and load them on every request.
 */
export type CacheType = 'preload' | 'expiry' | 'none';

export interface TemplateConfig {
  /**
   * Helper functions passed to handlebar templates
   */
  helpers?: RuntimeOptions['helpers'];
  /**
   * Main directory where handlebar files should be located, defaults to current working directory.
   * Looks for "partials" and "views" directories inside of it.
   */
  dir?: string;
  /**
   * Directory where all the partials are located, inside the main 'dir' directory, defaults to 'partials'
   */
  partials?: string;
  /**
   * Name of the handlebars file extension. Defaults to 'handlebars', so that only 'my-file.handlebars' kind of files are loaded.
   */
  extension?: string;
  /**
   * Directory name where all the views are located, inside the main 'dir' directory, defaults to 'views'
   */
  views?: string;
  /**
   * Cache types.
   * - preload: load all the templates ones into memory
   * - expiry: load partials when needed and keep them in cache, but check the file's mtime and reload template if changed
   * - none: do not cache templates and load them on every request.
   *
   * Defaults to 'preload'
   */
  cacheType?: CacheType;
  /**
   * Global data to be passed to every template
   */
  globalData?: Record<string, unknown>;
}

export interface CompileTemplatesOptions {
  dir: string;
  extension: string;
}

export type HandlebarsRender = (view: string, data?: Record<string, unknown>) => string;

export interface HandlebarsContext {
  hbs: HandlebarsRender;
}

const hasExtension =
  (extension: string) =>
  (file: string): boolean =>
    file.endsWith(extension);

export const deepReaddirSync = ({ dir, extension }: CompileTemplatesOptions): string[] =>
  readdirSync(dir, { withFileTypes: true })
    .reduce<string[]>((all, item) => {
      return all.concat(
        item.isDirectory() ? deepReaddirSync({ dir: join(dir, item.name), extension }) : join(dir, item.name),
      );
    }, [])
    .filter(hasExtension(extension));

const toName = (dir: string, file: string): string => relative(dir, file.slice(0, file.lastIndexOf('.')));

export const compileTemplates = ({ dir, extension }: CompileTemplatesOptions): Array<[string, TemplateDelegate]> =>
  deepReaddirSync({ dir, extension }).reduce<Array<[string, TemplateDelegate]>>(
    (all, file) => [...all, [toName(dir, file), compile(readFileSync(file, 'utf8'))]],
    [],
  );

export const loadTemplate = (
  name: string,
  cache: Cache,
  cacheType: CacheType,
  { dir, extension }: CompileTemplatesOptions,
): TemplateDelegate => {
  const file = `${join(dir, name)}.${extension}`;

  const stat = cacheType === 'expiry' ? statSync(file) : { mtime: new Date(0) };
  const cached = cache.get(name);

  if (cacheType !== 'none' && cached && stat.mtime < cached.mtime) {
    return cached.template;
  } else {
    const template = compile(readFileSync(file, 'utf8'));
    cache.set(name, { template, mtime: cacheType === 'expiry' ? stat.mtime : new Date() });
    return template;
  }
};

const loadPartials = (options: CompileTemplatesOptions): { [key: string]: TemplateDelegate } =>
  existsSync(options.dir)
    ? compileTemplates(options).reduce((acc, [name, template]) => ({ ...acc, [name]: template }), {})
    : {};

const loadTemplates = (cache: Cache, options: CompileTemplatesOptions): Cache =>
  existsSync(options.dir)
    ? compileTemplates(options).reduce(
        (cache, [name, template]) => cache.set(name, { template, mtime: new Date() }),
        cache,
      )
    : cache;

export const handlebars = ({
  helpers = {},
  dir = process.cwd(),
  extension = 'handlebars',
  partials = 'partials',
  views = 'views',
  cacheType = 'preload',
  globalData = {},
}: TemplateConfig): HandlebarsRender => {
  const cache = cacheType === 'preload' ? loadTemplates(new Map(), { dir, extension }) : new Map();
  const partialTemplates = cacheType === 'preload' ? loadPartials({ dir: join(dir, partials), extension }) : undefined;

  return (name: string, data: Record<string, unknown> = {}) => {
    const template = loadTemplate(name, cache, cacheType, { dir: join(dir, views), extension });
    return template(
      { ...globalData, ...data },
      { helpers, partials: partialTemplates ?? loadPartials({ dir: join(dir, partials), extension }) },
    );
  };
};

export const handlebarsMiddleware = (config: TemplateConfig): Middleware<HandlebarsContext> => {
  const hbs = handlebars(config);
  return (next) => (req) => next({ ...req, hbs });
};
