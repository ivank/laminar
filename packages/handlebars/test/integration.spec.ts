import { get, post, router, HttpService, run, htmlOk, htmlCreated } from '@laminar/laminar';
import axios, { AxiosResponse } from 'axios';
import { writeFileSync, mkdirSync } from 'fs';
import { retry } from 'ts-retry-promise';
import { join } from 'path';
import { handlebarsMiddleware } from '../src';

const axiosSnapshot = {
  config: expect.anything(),
  request: expect.anything(),
  headers: { date: expect.anything() },
};

const capitalizeName = (name: string): string => name.toUpperCase();

describe('Integration', () => {
  it('Should process response', async () => {
    const handlebars = handlebarsMiddleware({
      dir: join(__dirname, 'root'),
      helpers: { capitalizeName },
    });

    const http = new HttpService({
      port: 8062,
      listener: handlebars(
        router(
          get('/', async ({ hbs }) => htmlOk(hbs('index'))),
          post('/result', async ({ hbs, body: { name } }) => htmlCreated(hbs('result', { name }))),
        ),
      ),
    });

    const api = axios.create({ baseURL: 'http://localhost:8062' });

    await run({ initOrder: [http] }, async () => {
      expect(await api.get('/')).toMatchSnapshot<AxiosResponse>(axiosSnapshot);
      expect(await api.post('/result', { name: 'John Smith' })).toMatchSnapshot<AxiosResponse>(axiosSnapshot);
    });
  });

  jest.setTimeout(10000);

  it.each`
    cacheType    | expectedData
    ${'expiry'}  | ${'<html><body>Layout 2 <span>Generated 2</span></body></html>'}
    ${'none'}    | ${'<html><body>Layout 2 <span>Generated 2</span></body></html>'}
    ${'preload'} | ${'<html><body>Layout <span>Generated</span></body></html>'}
  `('Should process changing handlebars tempaltes with $cacheType', async ({ cacheType, expectedData }) => {
    const dir = join(__dirname, '__generated__/root');
    mkdirSync(join(dir, 'views'), { recursive: true });
    mkdirSync(join(dir, 'partials'), { recursive: true });
    writeFileSync(join(dir, 'views/index.hbs'), '{{#> layout }}<span>Generated</span>{{/layout}}');
    writeFileSync(join(dir, 'partials/layout.hbs'), '<html><body>Layout {{> @partial-block }}</body></html>');

    await new Promise((resolve) => setTimeout(resolve, 500));

    const handlebars = handlebarsMiddleware({ dir, extension: 'hbs', cacheType });

    const http = new HttpService({
      port: 8062,
      listener: handlebars(router(get('/', async ({ hbs }) => htmlOk(hbs('index'))))),
    });

    const api = axios.create({ baseURL: 'http://localhost:8062' });

    await run({ initOrder: [http] }, async () => {
      expect((await api.get('/')).data).toEqual('<html><body>Layout <span>Generated</span></body></html>');

      writeFileSync(join(dir, 'partials/layout.hbs'), '<html><body>Layout 2 {{> @partial-block }}</body></html>');
      writeFileSync(join(dir, 'views/index.hbs'), '{{#> layout }}<span>Generated 2</span>{{/layout}}');

      await retry(async () => expect((await api.get('/')).data).toEqual(expectedData), { delay: 500, retries: 10 });
    });
  });
});
