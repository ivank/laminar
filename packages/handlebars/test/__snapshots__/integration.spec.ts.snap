// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration Should process response 1`] = `
{
  "config": Anything,
  "data": "<html>
  <body>
    <span>Index</span>
  </body>
</html>
",
  "headers": {
    "connection": "close",
    "content-length": "57",
    "content-type": "text/html",
    "date": Anything,
  },
  "request": Anything,
  "status": 200,
  "statusText": "OK",
}
`;

exports[`Integration Should process response 2`] = `
{
  "config": Anything,
  "data": "<html>
  <body>
    <span>name: JOHN SMITH</span>
    <span><span>has test John Smith</span>
    </span>
  </body>
</html>
",
  "headers": {
    "connection": "close",
    "content-length": "123",
    "content-type": "text/html",
    "date": Anything,
  },
  "request": Anything,
  "status": 201,
  "statusText": "Created",
}
`;
